kernel/signal.c

```diff
@@ -97,13 +97,38 @@ void deliver_signal(struct task *task, int sig, struct siginfo_ info) {
     unlock(&task->sighand->lock);
 }
 
+void signal_delivery_stop(struct task *task, int sig, struct siginfo_ info) {
+    lock(&task->ptrace.stop_lock);
+    task->ptrace.stopped = true;
+    task->ptrace.signal = sig;
+    task->ptrace.info = info;
+    unlock(&task->ptrace.stop_lock);
+    notify(&current->parent->group->child_exit);
+    // TODO add siginfo
+    send_signal(current->parent, current->group->leader->exit_signal, SIGINFO_NIL);
+    unlock(&pids_lock);
+
+    lock(&current->ptrace.stop_lock);
+    while (current->ptrace.stopped) {
+        if (wait_for(&current->ptrace.stop_cond, &current->ptrace.stop_lock, NULL) == _EINTR)
+            receive_signals();
+    }
+    unlock(&current->ptrace.stop_lock);
+    lock(&pids_lock);
+}
+
 void send_signal(struct task *task, int sig, struct siginfo_ info) {
     // signal zero is for testing whether a process exists
     if (sig == 0)
         return;
     if (task->zombie)
         return;
 
+    if (task->traced && sig != SIGKILL_) {
+        signal_delivery_stop(task, sig, info);
```

Comment:
Shouldn't this be called from the receiving side of the signal, not the sending side? I think if you use `kill` on a ptraced process this would try to put the `kill` in signal-delivery stop.

Reply:
Uh, yes, I think it should. This might break a whole lot of stuff though so I'll have to look into it.

Reply:
Hopefully it also fixes a lot of stuff